ui/app/selectors/selectors.js

```diff
@@ -386,3 +444,135 @@ function getKnownMethodData (state, data) {
+  return Boolean(getFirstPermissionRequest(state))
+}
+
+function getPermissionsDomains (state) {
+  return state.metamask.domains
+}
+
+function getAddressConnectedDomainMap (state) {
+  const {
+    domains,
+    domainMetadata,
+  } = state.metamask
+
+  const addressConnectedIconMap = {}
+
+  if (domains) {
+    Object.keys(domains).forEach(domainKey => {
+      const { permissions } = domains[domainKey]
+      const { icon, name } = domainMetadata[domainKey]
+      permissions.forEach(perm => {
+        const caveats = perm.caveats || []
+        const exposedAccountCaveat = caveats.find(caveat => caveat.name === 'exposedAccounts')
+        if (exposedAccountCaveat && exposedAccountCaveat.value && exposedAccountCaveat.value.length) {
+          exposedAccountCaveat.value.forEach(address => {
+            addressConnectedIconMap[address] = addressConnectedIconMap[address]
+              ? { ...addressConnectedIconMap[address], [domainKey]: { icon, name } }
+              : { [domainKey]: { icon, name } }
+          })
+        }
+      })
+    })
+  }
+
+  return addressConnectedIconMap
+}
+
+function getDomainToConnectedAddressMap (state) {
+  const { domains = {} } = state.metamask
+
+  const domainToConnectedAddressMap = mapObjectValues(domains, (_, { permissions }) => {
+    const ethAccountsPermissions = permissions.filter(permission => permission.parentCapability === 'eth_accounts')
+    const ethAccountsPermissionsExposedAccountAddresses = ethAccountsPermissions.map(permission => {
+      const caveats = permission.caveats
+      const exposedAccountsCaveats = caveats.filter(caveat => caveat.name === 'exposedAccounts')
+      const exposedAccountsAddresses = exposedAccountsCaveats.map(caveat => caveat.value[0])
+      return exposedAccountsAddresses
+    })
+    const allAddressesConnectedToDomain = ethAccountsPermissionsExposedAccountAddresses.reduce((acc, arrayOfAddresses) => {
+      return [ ...acc, ...arrayOfAddresses ]
+    }, [])
+    return allAddressesConnectedToDomain
+  })
+
+  return domainToConnectedAddressMap
+}
+
+function getAddressConnectedToCurrentTab (state) {
+  const domainToConnectedAddressMap = getDomainToConnectedAddressMap(state)
+  const originOfCurrentTab = getOriginOfCurrentTab(state)
+  const addressesConnectedToCurrentTab = domainToConnectedAddressMap[originOfCurrentTab]
+  const addressConnectedToCurrentTab = addressesConnectedToCurrentTab && addressesConnectedToCurrentTab[0]
+  return addressConnectedToCurrentTab
+}
+
+function getRenderablePermissionsDomains (state) {
+  const {
+    domains = {},
+    domainMetadata,
+    permissionsHistory,
+    permissionsDescriptions,
+    selectedAddress,
+  } = state.metamask
+
+  const renderableDomains = Object.keys(domains).reduce((acc, domainKey) => {
+    const { permissions } = domains[domainKey]
+    const permissionsWithCaveatsForSelectedAddress = permissions.filter(perm => {
+      const caveats = perm.caveats || []
```

Comment:
Should we be looking specifically here at the `exposedAccounts` caveat here, as in `getDomainToConnectedAddressMap`? I'm not sure if it makes sense to draw any conclusions about how to interpret the value of a caveat without checking the type first.

Reply:
9aa364bbe